### ISSUE TITLE
Reproducible Builds

### ISSUE THREAD

**Issue by IzzySoft** [yuroyami/syncplay-mobile#105]
At IzzyOnDroid we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more in our repo](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Trying for yours, I was able to successfully generate the APK using below "build recipe", but the resulting APKs were not identical. Differences are only in `libc++_shared.so` for each of the ABIs, where their size differs by 800 kB for arm64 (yours is almost twice as large as the one from our build):

```diff
  -rw-r--r--  0.0 unx    10096 b-     4386 defN 1981-01-01 01:01:02 9734baa0 lib/arm64-v8a/libandroidx.graphics.path.so
- -rw-r--r--  0.0 unx  1822720 b-   508893 defN 1981-01-01 01:01:02 75660256 lib/arm64-v8a/libc++_shared.so
+ -rw-r--r--  0.0 unx  1027408 b-   305190 defN 1981-01-01 01:01:02 0d417b8c lib/arm64-v8a/libc++_shared.so
  -rw-r--r--  0.0 unx  2099504 b-   818001 defN 1981-01-01 01:01:02 91d16a3c lib/arm64-v8a/libconscrypt_jni.so
  -rw-r--r--  0.0 unx     7112 b-     2630 defN 1981-01-01 01:01:02 4f56ac38 lib/arm64-v8a/libdatastore_shared_counter.so
  -rw-r--r--  0.0 unx  1483648 b-   800769 defN 1981-01-01 01:01:02 3b0c45ae lib/arm64-v8a/libffmpegJNI.so
  -rw-r--r--  0.0 unx    80336 b-    22462 defN 1981-01-01 01:01:02 b01eb93b lib/arm64-v8a/libvlcjni.so
  -rw-r--r--  0.0 unx     7252 b-     4064 defN 1981-01-01 01:01:02 9a714d3a lib/armeabi-v7a/libandroidx.graphics.path.so
- -rw-r--r--  0.0 unx  1524360 b-   468153 defN 1981-01-01 01:01:02 81a90518 lib/armeabi-v7a/libc++_shared.so
+ -rw-r--r--  0.0 unx   610272 b-   235378 defN 1981-01-01 01:01:02 e4cdaf91 lib/armeabi-v7a/libc++_shared.so
  -rw-r--r--  0.0 unx  1279936 b-   637276 defN 1981-01-01 01:01:02 123cd601 lib/armeabi-v7a/libconscrypt_jni.so
  -rw-r--r--  0.0 unx     4416 b-     2033 defN 1981-01-01 01:01:02 ffa6438b lib/armeabi-v7a/libdatastore_shared_counter.so
  -rw-r--r--  0.0 unx  1436220 b-   791226 defN 1981-01-01 01:01:02 f367690d lib/armeabi-v7a/libffmpegJNI.so
  -rw-r--r--  0.0 unx    51328 b-    19202 defN 1981-01-01 01:01:02 3295e8ef lib/armeabi-v7a/libvlcjni.so
  -rw-r--r--  0.0 unx     9284 b-     4782 defN 1981-01-01 01:01:02 3d9afc66 lib/x86/libandroidx.graphics.path.so
- -rw-r--r--  0.0 unx  1602440 b-   505426 defN 1981-01-01 01:01:02 e390d8eb lib/x86/libc++_shared.so
+ -rw-r--r--  0.0 unx   993108 b-   336151 defN 1981-01-01 01:01:02 ae13de9f lib/x86/libc++_shared.so
  -rw-r--r--  0.0 unx  2254740 b-   904099 defN 1981-01-01 01:01:02 b728d7f5 lib/x86/libconscrypt_jni.so
  -rw-r--r--  0.0 unx     5148 b-     2330 defN 1981-01-01 01:01:02 db8d550b lib/x86/libdatastore_shared_counter.so
  -rw-r--r--  0.0 unx  1473020 b-   773162 defN 1981-01-01 01:01:02 41ced4ee lib/x86/libffmpegJNI.so
  -rw-r--r--  0.0 unx    76004 b-    25677 defN 1981-01-01 01:01:02 4dcddf64 lib/x86/libvlcjni.so
  -rw-r--r--  0.0 unx    10760 b-     4614 defN 1981-01-01 01:01:02 81bd88ae lib/x86_64/libandroidx.graphics.path.so
- -rw-r--r--  0.0 unx  1632144 b-   484338 defN 1981-01-01 01:01:02 27714799 lib/x86_64/libc++_shared.so
+ -rw-r--r--  0.0 unx  1045960 b-   330990 defN 1981-01-01 01:01:02 2d544d1c lib/x86_64/libc++_shared.so
  -rw-r--r--  0.0 unx  2501448 b-   967374 defN 1981-01-01 01:01:02 5e0ce8d6 lib/x86_64/libconscrypt_jni.so
```

We're building the exoOnly variant with [rbtlog](https://github.com/obfusk/rbtlog) on Debian Bookworm, using JDK 17. Here's the build recipe I've used:

```yaml
        build:
          - sed -r '/signingConfigs.getByName/d; s/val exoOnly = false/val exoOnly = true/' -i androidApp/build.gradle.kts
          - chmod +x gradlew
          - ./gradlew assembleRelease
          - mv androidApp/build/outputs/apk/noLibs/release/*unsigned.apk /outputs/unsigned.apk
        build_home_dir: /build
        build_repo_dir: /build/repo
        build_user: build
```

We'd appreciate if you could help making your build reproducible. We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that.

Looking forward to your reply!

cc @obfusk

**Comment by yuroyami** [comment]
@IzzySoft I reviewed the hints on reproducible builds and didn't find any false flags that could affect the build's reproducibility. The only concern is the size difference in the C++ shared library. I'm not entirely sure what's causing this size discrepancy but after checking out [C++ Library Support - Android NDK](https://developer.android.com/ndk/guides/cpp-support), I discovered that the `libc++_shared.so` library file's size depends on the NDK version used. It's also worth noting that the system can provide a "default" version of this library but I am not sure when that happens. What do you think ?

**Comment by IzzySoft** [comment]
I can specify the NDK to be used. AFAIK it should usually be discovered automatically, but in case a different one is needed (or the process somehow doesn't find the NDK), it can be specified. Currently, it's left on auto-discovery in the recipe for your app (so it found one, but that might be different from what you've used). If I understand that correctly, selection is then left to sdkmanager – I don't know how that picks its choice.

I'm not an Android dev, so  appreciate any "lessons" there: is there a standard way to specify the NDK version to be used, eg. in the build gradle? If so, I happily test it out if you tell me how (I could adjust the `build.gradle` in the build recipe here which would allow me to work with the current release – or we take an updated version from a different commit, but then we'd also need a matching APK).

I see that AGP allows to [specify `ndkPath`](https://developer.android.com/studio/projects/configure-agp-ndk?hl=en). But if you use a custom location there, I'd need to match that in my recipe as well requiring manual changes whenever you update the NDK. I can also specify the NDK version directly here, which would be the easiest way to test if this really is the culprit. Which exact NDK version were you using?

PS: I just found that one can [use something like `android.ndkVersion '19.2.5345600'` in `build.gradle`](https://stackoverflow.com/a/58133903/2533433). Just mentioning it here so in case the NDK version solves this, we remember this would probably be the preferred way (no "custom path", just the exact version). I'd expect sdkmanager to pick that up then.

-----

**TL;DR:** first step, if you tell me which exact NDK version you were using, I'd explicitly specify that and give it another try. If that solves the issue, we should try with `android.ndkVersion` in your `build.gradle`.

**Comment by IzzySoft** [comment]
OK, you didn't tell me – but as you've added it to your `build.gradle` I just tried to explicitly specify NDK 26.3.11579264. No dice, issue persists: `libc++_shared.so` has huge size differences. The same as above btw.

Not sure if it's related, but cannot hurt to ask: what do you use to compile on? We currently run the build on Debian bookworm with JDK 17.

**Comment by yuroyami** [comment]
@IzzySoft 
I specified the NDK version in the gradle build file but haven't re-released the APKs yet. I was planning to ping you once a new version is out so we can check everything. I'm also unsure if Android Studio used that exact version to build the files, as I have multiple NDK versions installed. Per any chance I assume it picks the latest one? 

Anyway, with the NDK version specified, if I release the APKs again and the issue persists, we can rule out an NDK mismatch. But since you've already noticed this, it seems unrelated to the NDK version. Not sure what's going on here.
I build on macOS Sonoma with JDK 17. I'll update you once I release a new version, just to confirm the NDK version mismatch isn't the problem.

**Comment by IzzySoft** [comment]
Ah, so my action was premature – apologies. I was just walking the backlog and happened to reach your app…

> Anyway, with the NDK version specified, if I release the APKs again and the issue persists, we can rule out an NDK mismatch. But since you've already noticed this, it seems unrelated to the NDK version.

Not necessarily– unless the APKs currently attached have been built with exactly that NDK. You can provie me an APK here (just rename it to `*.zip` so Github allows attaching it, don't put it into another zip – that way I can run it directly from here) and name the matching commit. I then can see if RB succeeds with that.

But yeah, I can also wait for the next release if you prefer.

**Comment by IzzySoft** [comment]
Btw, as a heads-up: I've just tried v0.15.0 again with explicitly specifying `ndk: 26.3.11579264` and building on `ubuntu:jammy`. I get exactly the same APK (hash matches).

**Comment by IzzySoft** [comment]
Just wondering where we stand here, @yuroyami ? Long time not heard a thing. If we're both out of ideas, I'd have a last: would it be an option to use Github actions to build the app? Then I might be able to look up what you define there and try to match it with my build recipe. Going by the sizes in the diff, it looks like something is missing here at my end:

```diff
- -rw-r--r--  0.0 unx  1822720 b-   508893 defN 1981-01-01 01:01:02 75660256 lib/arm64-v8a/libc++_shared.so
+ -rw-r--r--  0.0 unx  1027408 b-   305190 defN 1981-01-01 01:01:02 0d417b8c lib/arm64-v8a/libc++_shared.so
```

That's a difference of ~800 kB uncompressed. So something is missing here. I simply run `./gradlew assembleRelease` – are there some steps needed before that? (I cannot use GUI clients for this, so I need to build directly with gradle from the command line; build instructions in your Readme are just for Android Studio. This is one of the reasons I suggested Github actions, so you'd see for yourself).

**Comment by yuroyami** [comment]
@IzzySoft I think some third-party steps are involved in the build. Since this is a KMP project, I’m not exactly sure how the build process compares to pure Android projects, especially with the Kotlin Multiplatform plugin in play, but hopefully, that’s not the issue and the problem is something else. I’m also not sure if GitHub Actions can build a KMP project, but I believe it should be doable. I haven’t had a chance to work on this project again, but I hope to pick it up later this month and see if I can deploy builds through GitHub Actions. I’ll ping you when it’s ready.

**Comment by IzzySoft** [comment]
> I’m also not sure if GitHub Actions can build a KMP project

If they can't, I guess our builders would not be able either. But we'll see.

> I hope to pick it up later this month and see if I can deploy builds through GitHub Actions. I’ll ping you when it’s ready.

Thanks! Fingers crossed then – and looking forward to your ping!

**Comment by IzzySoft** [comment]
Just a friendly ping (no pressure): any success yet? If you have a Linux machine to build on, we meanwhile also could offer you our [rbuilder_setup](https://codeberg.org/IzzyOnDroid/rbuilder_setup) scripts to set up a builder like the ones we use in just 5 minutes. Plus the recipe we've used for SyncPlay so far, to get you started. While intended to confirm builds, cross-checking against another APK, they can also be used to "simply build an APK". I'm pretty sure the results would be RB then 😉

**Comment by yuroyami** [comment]
@IzzySoft I am currently on macOS but I also have another environment with Windows so I could run it on WSL. I will try it out, thanks!

**Comment by IzzySoft** [comment]
Wonderful! I'm very interested to hear how this works in WSL. Since yesterday, rbuilder_setup also supports RPM and Arch based systems (Readme has just been brought up-to-date a few minutes ago) – but I guess you'll go with some Ubuntu there?

**Comment by IzzySoft** [comment]
Any news there, @yuroyami? I'm double curious now. First, to get Syncplay RB – but second also about how our framework works on WSL 😃

**Comment by yuroyami** [comment]
@IzzySoft  Sorry for the delay! I've been super swamped these past few months and only recently got the chance to revisit this project. I'm currently making major changes to its overall structure and expect to release a new version within a month or so. I'll likely have GitHub Actions set up by then as well.

I didn’t get a chance to test the script on WSL Ubuntu (ran into storage issues and had to drop that setup shortly after my last comment). At the moment, I’m working solely on a Mac. We can try out the GitHub Actions thing first and see where we can go from there.

**Comment by IzzySoft** [comment]
Thanks for the feedback! Yeah, I'm permanently "swamped", so I can relate… And yes, GH actions is fine. Would just have been interesting to know our framework running on WSL, too. But of course, whatever suits you best. Just let me know when the new APK is ready for testing then 😉

**Comment by yuroyami** [comment]
@IzzySoft Ops, my bad! My apologies. This entire Gradle task is not supposed to run for the exoOnly flavor. It's only designed to download and build dependencies for mpv's native libs. The release that gets shipped to IzzyDroid is an exo-only release without any native libs, therefore the whole task is to be skipped. I will make the necessary adjustments and ping you again

**Comment by IzzySoft** [comment]
We got another issue there. Our updater just fetched the APK, and yielded

```
2025-11-19 20:04:35,983 WARNING: "com.reddnek.syncplay_1000016000.apk" is signed by a key that is not allowed:
cfeed0f6458db903ef44f558ed97a6edd0a656bb76bc645178b3f266de2a9566
```

What happened to your signing key? We now have to disable updates for SyncPlay until this is solved.

**Comment by yuroyami** [comment]
@IzzySoft I lost my signing key which I also use for Google Play store uploads due to an unfortunate SSD failure. The new APK is signed with a new one. Sorry for the inconvenience.
I am also thinking of changing the appId from `com.reddnek.syncplay` to `com.yuroyami.syncplay` as i don't see why i should hold on to the old one anymore. I can generate a new exo-only APK with that change as well

**Comment by IzzySoft** [comment]
> I lost my signing key

That is very unfortunate. Now, how can we verify it's really you – and not someone who managed to take over your Github account? You left no alternative way of contacting you – and you did not sign your commits with your own key (see [Sign your commits](https://izzyondroid.org/docs/devpractices/SignCommits/) in our documentation); the only signed commits I see are signed by Github with their key. So, please take a look at [How to keep your key safe and what measures to take for the event of loss?](https://f-droid.org/2023/09/03/reproducible-builds-signing-keys-and-binary-repos.html#lessons-learned-2-how-to-keep-your-key-safe-and-what-measures-to-take-for-the-event-of-loss) Then let us know what path of verification you see can be used.

> I am also thinking of changing the appId from `com.reddnek.syncplay` to `com.yuroyami.syncplay` as i don't see why i should hold on to the old one anymore.

Because only "holding on to it" ensures that people can update? Let me quote from the history of this app:

```yaml
  Status:
    - 2022-07-10 rebranded from com.gintoki.syncplay which was originally added 2022-05-06
    - 2022-08-31 rebranded again from com.chromaticnoob.syncplay (should it happen once more this year, we better drop the app from the repo) https://github.com/Reddnek/syncplay-android/issues/35
    - 2023-01-31 rebranded again from com.reddnek.syncplay – sent a reminder to the issue this will be the last time. No grace period this time, we simply rename here as well.
    - 2023-02-01 rebranding to com.reddnek.syncplay.revamped reversed
```

And then, let me quote [from the linked issue](https://github.com/yuroyami/syncplay-mobile/issues/35#issuecomment-1410934306):

> When will you stop renaming the package? That's now the 3rd time in just 6 months. This not only causes a bunch of extra work on my end, but also forces your users to uninstall and re-install the app each time. To be honest, I'm tired of cleaning up every time.
>
> This is the last time I follow your rename, but without keeping the previous version for an overlap. Next time I might simply drop it. I run my repo for more than 6 years now and your app is the only one behaving like this; a few had a single rename, but none 3 within a year – if it continues this way, it's rather messy for all involved.

I hope that explains "why"? I maintain this repo for about 10 years now, and the above still holds true.

**Comment by IzzySoft** [comment]
> and it took weeks to fix because Google support was hell to deal with.

And now it's taking months here to fix because… ? As I wrote 2 month ago, we'll now disable update checks entirely.

**Comment by yuroyami** [comment]
> And now it's taking months here to fix because… ? As I wrote 2 month ago, we'll now disable update checks entirely.

I had replied to you via email, but it seems my response didn’t reach you.
I’ve since transferred the app from my friend’s Play Console account to my own so I could complete the keystore change process more easily, which I’ve now done. The keystore has been updated on Google Play and should already be in effect. 

Edit: I suppose the next step is I push a new versionto GPlay so you can match its key signature with the APK I upload here in GitHub?

**Comment by yuroyami** [comment]
> To which address have you sent a mail?

I received a bounce-back (“renvoi”) notification indicating that my email wasn’t delivered. So it likely never reached you.
In that message, I mentioned that I’m open to any ownership verification method. The only technically viable approach I could think of was verifying the Google Play signing certificate mismatch via the keystore hash.

>  I hope you don't use their signing?

I am using Google Play App Signing. I overlooked the distinction between the upload keystore and the app signing key. Since Google re-signs the distributed APKs with their app signing key, my earlier keystore-hash idea wouldn’t work in this case.

> It's not that they offer download links or such.

Technically, an installed app can be extracted and re-packaged as an APK. However, I’m not certain whether the original signing block is preserved post-installation, my assumption is that it’s stripped during the install process (but only guessing).

So keeping Google Play tricksies aside, what do you suggest we do next? I made more attempts trying to recall the keystore password but no avail...

**Comment by IzzySoft** [comment]
Well, fat chance with that. Quoting from the build logs:

```
> Configure project :shared
?? Version fields not found in project.pbxproj
w: ?? Unsupported Operating System
Kotlin CocoaPods Plugin is fully supported on MacOS machines only. Gradle tasks that can not run on non-mac hosts will be skipped.
Solution:
Run the build on a MacOS machine

...

sh mpv_build.sh -n syncplay-withmpv
./include/download-sdk.sh: line 16: sudo: command not found
```

Requiring `sudo` is a killer, that won't ever be available inside the build recipe, sorry. Packages must be installed during provisioning, not during build. And it would be good to have a way to tell the build to only process the targets needed (here: Android).

That said, I've managed to build by installing the requirements during provisioning, with this recipe:

```yaml
build:
  - git reset --soft 2c5e343463c7506e799bd5292aeb7176fe5502a9
  - sed -r 's/const val exoOnly = false/const val exoOnly = true/' -i buildSrc/src/main/kotlin/AppConfig.kt
  - ./gradlew assembleRelease
```

but the resulting APK was far from identical (diffoscope produced a 40+ MB HTML pointing out the differences). First, the APK was not built from a clean tree at the tagged commit – but either from the commit before that, or with the changes of the tagged commit present but uncommitted; hence the `git reset`. Then I had to somehow convince the build to produce the `exoOnly` APK – I found no flavors defined, so I helped myself with that `sed` (guess there's an `assembleExoOnly` for that).

Attaching you all the diffs and the recipe here, so you can look for yourself; I might have missed things there, so if you can tell me what, I can try again.

[diff.zip](https://github.com/user-attachments/files/26607667/diff.zip)

**Comment by IzzySoft** [comment]
@yuroyami I see a new release, but I also see your scripts still include the `sudo` calls. Any plans to solve that – or do we have to drop all attempts to confirm RB for Syncplay?

**Comment by yuroyami** [comment]
@IzzySoft I may be able to drop the sudo calls in the native scripts with help of a coding agent as I'm in no way proficient in C. I'll check it out today

**Comment by IzzySoft** [comment]
Careful with LLMs, see our [AI policy](https://izzyondroid.org/docs/general/AppInclusionPolicy/#ai-policy). The `sudo` commands are "build requirements", and such should go to provisioning (e.g. in your Github action), like

```yaml
- name: install build dependencies
  run: sudo apt-get install autoconf pkg-config libtool ninja-build unzip wget meson
```

**Comment by yuroyami** [comment]
@IzzySoft Update on the build side: the exo-only variant is now selectable from the command line, so you no longer have to patch `AppConfig.kt` by hand.

It's a Gradle property now (in the top-level `gradle.properties`). To build the APK shipped to IzzyOnDroid:

```
./gradlew assembleRelease -PexoOnly=true
```

All the native-script tasks are guarded by the `exoOnly` flag, so when it's `true` they don't even get created in the first place. The reason your recipe kept stumbling on the `sudo` calls and native scripts is that the build was resolving to the full flavor, not the exo-only one. So your recipe's build step should reduce to roughly:

```yaml
build:
  - ./gradlew assembleRelease -PexoOnly=true
```

If you'd also like the `sudo` cleanup on top, just say so, though I don't think it's necessary at this point.

Now back to the diff you uploaded earlier in this thread. I went through it, and good news first: the build is *almost* fully reproducible. Out of 881 entries, only two things differ: `libc++_shared.so` (all 4 ABIs) and `classes.dex` / `classes2.dex` (plus the two `baseline.prof*` files). Everything else, including all of `resources.arsc`, matches byte-for-byte.

1. `libc++_shared.so`: I traced this one carefully this time. My full build (mpv/VLC) drops a `libc++_shared.so` byproduct into `src/main/libs/<abi>/`, which is gitignored. On my machine that leftover was present and got packaged into the release; on your clean checkout it doesn't exist, so your build pulled the **VLC AAR's** `libc++_shared.so` instead (a different binary), hence the mismatch. So your side was right all along; my published APK was the odd one out. I've changed the exo-only build to prune that stray `libc++_shared.so` before packaging, so it now **deterministically** uses the VLC AAR's copy, the same one your clean build picks. That should make all 4 ABIs match on the next release.

2. `classes.dex` / `classes2.dex`: the compiled code differs, which usually points at the toolchain. Two things in the recipe stand out:
   - It provisions `openjdk-17`, but the project builds with a **JDK 21** toolchain (`org.gradle.toolchains.jvm.version=21`). Could you bump the recipe to JDK 21? I think this is the culprit.
   - `build_tools:` is left blank, so it isn't pinned. I've now pinned it to `36.0.0` on my side. It probably isn't the cause though, since `resources.arsc` (which build-tools produces) already matches byte-for-byte, so don't expect much from this one. JDK 21 is the priority.

The two `baseline.prof*` files are generated from the dex, so they should fall in line once the dex matches.

I'll cut a new release (v0.22.0, you'll see it within 30 minutes) with the libc++ fix and the `-PexoOnly=true` flag from above. If you rebuild with JDK 21 + pinned build-tools against it, we should be a lot closer, hopefully all the way.

**Comment by IzzySoft** [comment]
Pulling stuff from Github is not as reliable as it once has been…

```
Submodule 'dlg' (https://github.com/nyorain/dlg.git) registered for path 'subprojects/dlg'
Cloning into '/build/repo/buildscripts/deps/freetype2/subprojects/dlg'...
Submodule path 'subprojects/dlg': checked out '395ccad2c1e0daae535c4d20bb0a3f2424648e17'
tar (grandchild): xz: Cannot exec: No such file or directory
tar (grandchild): Error is not recoverable: exiting now
--2026-05-25 09:15:30--  https://github.com/fribidi/fribidi/releases/download/v1.0.16/fribidi-1.0.16.tar.xz
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://release-assets.githubusercontent.com/github-production-release-asset/5575048/92060972-e764-4abf-9cc0-ab2f06fe4b62?sp=r&sv=2018-11-09&sr=b&spr=https&se=2026-05-25T10%3A09%3A02Z&rscd=attachment%3B+filename%3Dfribidi-1.0.16.tar.xz&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2026-05-25T09%3A08%3A29Z&ske=2026-05-25T10%3A09%3A02Z&sks=b&skv=2018-11-09&sig=kqnxS%2FU2gpXzBBlFTGHtBXkxBsr%2B%2B4dc1pQlCxWO0p0%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc3OTcwMDgzMCwibmJmIjoxNzc5NzAwNTMwLCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvbi5ibG9iLmNvcmUud2luZG93cy5uZXQifQ.gHIexQRzsd7QIzGpt2BVOY6IVeKtPepxoFf4cGgSlYQ&response-content-disposition=attachment%3B%20filename%3Dfribidi-1.0.16.tar.xz&response-content-type=application%2Foctet-stream [following]
--2026-05-25 09:15:30--  https://release-assets.githubusercontent.com/github-production-release-asset/5575048/92060972-e764-4abf-9cc0-ab2f06fe4b62?sp=r&sv=2018-11-09&sr=b&spr=https&se=2026-05-25T10%3A09%3A02Z&rscd=attachment%3B+filename%3Dfribidi-1.0.16.tar.xz&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2026-05-25T09%3A08%3A29Z&ske=2026-05-25T10%3A09%3A02Z&sks=b&skv=2018-11-09&sig=kqnxS%2FU2gpXzBBlFTGHtBXkxBsr%2B%2B4dc1pQlCxWO0p0%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc3OTcwMDgzMCwibmJmIjoxNzc5NzAwNTMwLCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvbi5ibG9iLmNvcmUud2luZG93cy5uZXQifQ.gHIexQRzsd7QIzGpt2BVOY6IVeKtPepxoFf4cGgSlYQ&response-content-disposition=attachment%3B%20filename%3Dfribidi-1.0.16.tar.xz&response-content-type=application%2Foctet-stream
Resolving release-assets.githubusercontent.com (release-assets.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to release-assets.githubusercontent.com (release-assets.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1098260 (1.0M) [application/octet-stream]
Saving to: 'STDOUT'

     0K .....tar: Child died with signal 13
tar: Error is not recoverable: exiting now
...                                                0% 9.07M=0.001s



Cannot write to '-' (Broken pipe).
> Task :androidApp:runAndroidMpvNativeBuildScripts FAILED
```

OK, on a second look, that rather seems to be a missing dependency (`xz-utils`), so let's try again:

```
Cloning into 'dav1d'...
fatal: unable to access 'https://github.com/videolan/dav1d/': Could not resolve host: github.com
```

We have 20+ failed update checks each day because of such things. GH is simply no longer reliable. They once have been at 99.99% (before MS took over) – nowadays they are below 85%. So while the build is running for a 3rd run, which of the dependencies are still needed, and which might be missing?

- autoconf
- pkg-config
- libtool
- ninja-build
- bzip2
- xz-utils

Third round got over that, but in the logs I see:

```
> Task :androidApp:runAndroidMpvNativeBuildScripts FAILED
mpv_build.sh: 3: Bad substitution
mpv_build.sh: 23: ./include/depinfo.sh: Syntax error: "(" unexpected
```

so build fails:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':androidApp:runAndroidMpvNativeBuildScripts' (registered in build file 'androidApp/build.gradle.kts').
> Process 'command 'sh'' finished with non-zero exit value 2
```

Strange: the error is reported for `sh`, while the script clearly states to use Bash (shebang for `#!/bin/bash`). And for Bash, the syntax is correct. `registered in build file 'androidApp/build.gradle.kts'`, huh?

```kotlin
if (!exoOnly) {
    tasks.named("preBuild") {
        dependsOn("runAndroidMpvNativeBuildScripts")
    }
}
```

Build was run with `./gradlew assembleRelease -PexoOnly=true`, so that block should not have executed. You're sure the flag is evaluated properly? [I don't think so](https://github.com/yuroyami/syncplay-mobile/blob/v0.21.10/buildSrc/src/main/kotlin/AppConfig.kt#L30). It's a constant, set to `false` – and you cannot change the value of a constant, once declared. So I'm afraid the `-PexoOnly=true` passed to gradlew, will have no effect. And "oh wonder": putting in the `sed` again, build succeeds. Still, not RB:

```diff
  -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 5628f820 META-INF/version-control-info.textproto
- -rw-r--r--  0.0 unx    17448 b-    17448 stor 1981-01-01 01:01:02 98bfa1d8 assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx    17449 b-    17449 stor 1981-01-01 01:01:02 d8b0bbf4 assets/dexopt/baseline.prof
  -rw-r--r--  0.0 unx     1990 b-     1990 stor 1981-01-01 01:01:02 53e2b45f assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  8333400 b-  3865710 defN 1981-01-01 01:01:02 b391c11c classes.dex
+ -rw-r--r--  0.0 unx  8332800 b-  3865583 defN 1981-01-01 01:01:02 e69b3a9b classes.dex
  -rw-r--r--  0.0 unx     3548 b-     1787 defN 1981-01-01 01:01:02 b41dc5af classes2.dex
```

Funny diff in the dex now. Are there any secrets involved? Like, an API key for … Klipy? Ah! Now! Including your "secret" does the trick then 🥳 

**TL;DR:**

* GitHub has become a mess (nothing you or I can fix – it's their "agentic workflows" eating their resources, so rather expect it to get worse)
* passing `-PexoOnly=true` has no effect, as you defined `exoOnly` as constant with the value `false`
* dependencies need to be checked: are they all needed still? I've left them in our recipe for now.
* we need to inject your "secret" (so hopefully that does not change too often). As usual, we "camouflage" that, of course, to make it not too obvious. That it can be easily retrieved, is already confirmed by the fact that I was able to – no special tools needed

Finally, [here comes your welcome wagon](https://floss.social/@IzzyOnDroid/116634817555339599) 🥳 

<img width="347" height="638" alt="Image" src="https://github.com/user-attachments/assets/3a648d55-082b-4481-95d2-99c2d6c8fefd" />

Thanks for your help with this! 🤗

**Comment by IzzySoft** [comment]
As this is still open: RB failed for v0.22.1 – please remove the Foojay resolver again, and do **NOT** pin the toolchain to a specific vendor:

* https://github.com/yuroyami/syncplay-mobile/blob/master/settings.gradle.kts#L23
* https://github.com/yuroyami/syncplay-mobile/blob/master/gradle/gradle-daemon-jvm.properties#L2C110-L12C25

Because:

```
Starting a Gradle Daemon (subsequent builds will be faster)                                                                                              
                                                                                                                                                         
FAILURE: Build failed with an exception.                                                                                                                 
                                                                                                                                                         
* What went wrong:                                                                                                                                       
Unable to download toolchain matching the requirements ({languageVersion=21, vendor=Eclipse Temurin, implementation=vendor-specific, nativeImageCapable=false}) from 'https://api.foojay.io/disco/v3.0/ids/ecd23fd7707c683afbcd6052998cb6a9/redirect', due to: Toolchain provisioned from 'https://api.foojay.io/disco/v3.0/ids/ecd23fd7707c683afbcd6052998cb6a9/redirect' doesn't satisfy the specification: {languageVersion=21, vendor=Eclipse Temurin, implementation=vendor-specific, nativeImageCapable=false} and must have the executable 'javac', the executable 'javadoc', and the executable 'jar'.                      
```

We were able to build (and succeeded with RB) by preceeding the recipe with

```yaml
- sed -r '/(foojay|toolchainVendor)/d' -i gradle/gradle-daemon-jvm.properties
- sed -r '/org.gradle.toolchains.foojay-resolver-convention/d' settings.gradle.kts
```

(i.e. doing exactly that: removing Foojay, and the pin to Adoptium).

**Comment by yuroyami** [comment]
It should be fixed now, can you check ? @IzzySoft
